### PR TITLE
[Enterprise] Add operations manual

### PR DESCRIPTION
**Pull request by schultyy**
This PR adds a new document which we call the operations manual. The purpose of this document is to provide guidance for our customers in case something goes wrong.

### CODE REVIEW

File: user/enterprise/operations-manual.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+- Your currently used Docker version needs to be at least 1.7 on both platform and worker machines.
+- For the platform machine, these port bindings have to be in place (run `sudo docker ps -a` and look for a container whose name starts with `quay.io-travisci-te-main`):
+	- 0.0.0.0:80->80/tcp
+	- 0.0.0.0:443->443/tcp
+	- 0.0.0.0:4567->4567/tcp
+	- 0.0.0.0:5672->5672/tcp
+	- 0.0.0.0:32770->22/tcp
+- If the above port bindings are not there, please let us know. Then we'll figure out how to get them back.
+- An important topic is the used storage driver. In the past we've heavily relied on the `devicemapper` storage driver. At some point Docker changed their default storage driver to `overlay`. However this only works well if your Linux kernel is at least version 3.18. Below it *can* work but there's no definite guarantee. To find out the used storage driver, please run `sudo docker info | grep "Storage Driver"`. If your current driver is `overlay`, please check your Linux kernel version via `uname -a`.
```
Comment (bnferguson) [inline user/enterprise/operations-manual.md]: This isn't quite the case. At least with the worker, we default to `aufs`, not `overlay` in Ubuntu. Xenial will be the first time we can use `overlay`. So it's possible they may see overlay on RedHat or Xenial but not on Trusty AFAIK. DeviceMapper setups require a lot of setup. So those are pretty deliberate.

File: user/enterprise/operations-manual.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+## Worker
+The `travis-worker` binary on the worker machine(s) must be up to date.
+Please run `sudo apt-get install --only-upgrade travis-worker`. This will update it to the latest version.
+
+__Docker__:
+
+Since the worker machine runs all your builds in Docker containers, the Docker version must be up to date and properly configured. We also need to have the `devicemapper` or `overlay` storage driver configured here, together with Linux kernel 3.18 or later.
```
Comment (bnferguson) [inline user/enterprise/operations-manual.md]: Like the platform machines, `aufs` is our default storage driver still. Though there are some `devicemapper` installs out there from our installer. I think only verizon is using Overlay with their customer worker setup.

File: user/enterprise/operations-manual.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+Please also check the worker configuration in `/etc/default/travis-enterprise`:
+
+- Please make sure that `TRAVIS_ENTERPRISE_BUILD_ENDPOINT` is not commented out and has the value `__build__`.
+- Also make sure that `TRAVIS_ENTERPRISE_SECURITY_TOKEN` has the same value as being shown under RabbitMQ password at https://<your domain>:8800/settings
```
Comment (bnferguson) [inline user/enterprise/operations-manual.md]: the `<your domain>` seems to get edited out since it's seeing it as markup. May need to put backticks around it.

File: user/enterprise/operations-manual.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
+---
+title: Travis CI Enterprise Operations manual
+layout: en_enterprise
+
+---
+
+This document is meant as a guideline to follow in case your Travis CI Enterprise production installation becomes unresponsive or unavailable.
```
Comment (plaindocs) [inline user/enterprise/operations-manual.md]: I'd say something like "This document contains guidelines to follow" ...

File: user/enterprise/operations-manual.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+If this happens, we'd like to ask you to follow the steps outlined in this document first before you get in touch with us to save important data and collect as much evidence as possible. This will help us to provide you the best support possible and focus completely on resolving the problem.
```
Comment (plaindocs) [inline user/enterprise/operations-manual.md]: And here I'd suggest being more direct:

> If your Travis CI Enterprise production installation becomes unresponsive, please follow the steps outlined in this document before contacting us, the data you collect helps us solve your problem...

File: user/enterprise/operations-manual.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+## First: Don't panic
```
Comment (plaindocs) [inline user/enterprise/operations-manual.md]: PERSONALLY I'd find this first step patronizing, but ymmv

If we're talking about them as steps, they should be numbered, as well.

File: user/enterprise/operations-manual.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+The first thing you should do is: keep calm. We're here to help and we're confident that we can bring it back online together with you!
+
+## Prerequisites
+
```
Comment (plaindocs) [inline user/enterprise/operations-manual.md]: This isnt a step, and also won't help when you get to this point. Fold it into the sentence on line 9 maybe, and probably shorten it a bit as well

File: user/enterprise/operations-manual.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+Please make sure that you have root access on all machines Travis CI Enterprise is running on. This is crucial because we will ask you to perform operations where you either need to be root or you need to have the appropriate `sudo` privileges. If you don't have direct root privileges, please make sure that a person who has is available.
+
+## Coordinating with Support
+
+We are here to help troubleshoot and fix any problems that may arise with your Travis CI Enterprise installation. For efficiency, though, we would request you minimize making changes to your setup that are not directly recommended by support, or document these steps and notify support accordingly.
```
Comment (plaindocs) [inline user/enterprise/operations-manual.md]: This is seems not really relevant given the specified purpose of the document

File: user/enterprise/operations-manual.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+## Terminology
```
Comment (plaindocs) [inline user/enterprise/operations-manual.md]: I'd probably skip this when reading a troubleshooting doc so I'd ditch the title. And I'd reword to something like

> Throughout this document we'll be using the following terms to refer to the two components of your Travis CI Enterprise installation:

- `Platform machine`: The virtual machine that runs most of the Travis web components. This is the machine your domain is pointing to.
- `Worker machine`: The worker machine(s) that run your builds.

In some parts of the docs we call these hosts? And in some machines? Can we be consistent

File: user/enterprise/operations-manual.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+To make sure that there are no confusions, we'd like to define a couple of terms:
+
+- `Platform machine`: The machine that runs most of the Travis web components. This is the machine your domain is pointing to.
+- `Worker machine`: The worker machine(s) run your builds.
+
+## Backup your encryption key
```
Comment (plaindocs) [inline user/enterprise/operations-manual.md]: "Back up"  when used as a verb

"very important building block " doesn't add much, so it becomes 

Without the encryption key you cannot access the information in your production database. To make sure that you can always recover your database, make a backup of this key:
Make this a note:

> Without this key the information in the database is not recoverable.

File: user/enterprise/operations-manual.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+This key is a very important building block in Travis CI Enterprise. It is used to encrypt certain parts of your production database. _Without this key the information in the database is not recoverable._ To make sure that the database remains accessible we ask you to make a backup of this key by following the steps below:
+
+- open a ssh connection to the platform machine
```
Comment (plaindocs) [inline user/enterprise/operations-manual.md]: Number these steps

File: user/enterprise/operations-manual.md [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+- run `travis bash`. This will open a bash session with `root` privileges into the Travis container.
+- Then run `cat /usr/local/travis/etc/travis/config/travis.yml | grep -A1 encryption:`. Create a backup of value getting printed on the screen by either  writing it down on a piece of paper or storing it on a different computer.
+
+## Create a backup of the data directories
+
+The data directories are located on the platform machine and get mounted into the Travis container. In there you'll find files from RabbitMQ, Postgres, Slanger and Postgresql.
```
Comment (plaindocs) [inline user/enterprise/operations-manual.md]: "In there" refers to the container, which is neither the platform mahcine nor the worker machine, which is confusing given those are two things we're talking about.

File: user/enterprise/operations-manual.md [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+These files are located at `/var/travis`. Please run `sudo tar -czvf travis-enterprise-data-backup.tar.gz /var/travis` to create compressed archive from this folder. After this has finished, copy this file off the machine to a secure location.
+
+## Pull a Support bundle
+
+Please head over to https://<your domain>:8800/support and create a support bundle. If you're writing us, please include that file. It'll include all important log files which will help us to diagnose the problem.
+
+Also please include the worker log files as well. They're located on your worker machine(s) at `/var/log/upstart/travis-worker.log`. If you run multiple worker machines, please grab this file from all machines.
+
+## What exactly is not running?
```
Comment (plaindocs) [inline user/enterprise/operations-manual.md]: Call this something like

## What information should I include in my support request? 

or even better, if it is a numbere step:

## Write a support request

File: user/enterprise/operations-manual.md [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+__Platform machine__:
+
+- Port 22 for SSH access
+- Port 443 for HTTPS traffic
+- Port 80 for HTTP traffic
+- Port 4567 & 5672 for RabbitMQ traffic
+- Port 8800 for management dashboard https traffic
+
+## Hardware
+
```
Comment (plaindocs) [inline user/enterprise/operations-manual.md]: Should link these in not repeat them?

File: user/enterprise/operations-manual.md [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+Our platform machine has these hardware requirements:
+
+- Minimum 16GB of RAM
+- 8 CPU cores
+- Min. 40GB of available disk space
+
+When you run `top` on the platform machine, please check if enough RAM is available and that the system load is not too high.
```
Comment (plaindocs) [inline user/enterprise/operations-manual.md]: Enough ram and load not too high is not specific. People might not know systems well enough to answer this

File: user/enterprise/operations-manual.md [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+Did you notice an increased system load recently?
+
+## Docker
+
+We use a tool called [Replicated](https://www.replicated.com/) to distribute Travis CI Enterprise. It also provides the management dashboard you're using to configure certain things.
```
Comment (plaindocs) [inline user/enterprise/operations-manual.md]: s/certain things/your installation/  ?

File: user/enterprise/operations-manual.md (hunk 13 above)
Comment (plaindocs) [reply user/enterprise/operations-manual.md]: And then every subsequent header should be `###`

File: user/enterprise/operations-manual.md (hunk 6 above)
Comment (plaindocs) [reply user/enterprise/operations-manual.md]: The only ones that are numberd steps are 

1. Back up your key
2. Back up your data directories
3. Pull a support bundle (can we call this "create" not "pull")
4. Write a support request

Comment (acnagy) [reply user/enterprise/operations-manual.md]: Hmm... actually, I think I agree with @plaindocs. What if we moved the sentence into "coordinating with support" - that way it becomes more of a statement about support working with you, etc. 

The thing is, I know what customer experiences have made us decide that a reassurance is necessary, _but_, I agree - better to build it in as encouragement rather than a "stop what you're doing" message

File: user/enterprise/operations-manual.md (hunk 7 above)
Comment (acnagy) [reply user/enterprise/operations-manual.md]: I agree with the rationale, but I think we should put this step after line 21 - it's about support, after all

File: user/enterprise/operations-manual.md (hunk 8 above)
Comment (acnagy) [reply user/enterprise/operations-manual.md]: I know what @schultyy's going for, and I think the concept (please coordinate with support if there's an issue, and if you are, document changes as you go) is pretty important... but, the fact it's not obvious is definitely an issue... 

Another idea: What if all of the stuff about support appeared with the support bundle discussion? That's really the only place that it matters - if the customer's gonna get in contact, we're gonna (most likely) need a support bundle, so the steps on how to coordinate with support are moot if that's not part of the troubleshooting

File: user/enterprise/operations-manual.md (hunk 9 above)
Comment (acnagy) [reply user/enterprise/operations-manual.md]: I like the definitions - people do use 'host' and 'platform' OR 'worker' interchangeably, and often expect master/slave diction too... however, perhaps this section can be moved somewhere where it's ~less likely to be skipped~ more helpful?

File: user/enterprise/operations-manual.md (hunk 13 above)
Comment (acnagy) [reply user/enterprise/operations-manual.md]: I like this idea! And, a final step can be "please document any changes you make while working with support...", etc

File: user/enterprise/operations-manual.md (hunk 15 above)
Comment (acnagy) [reply user/enterprise/operations-manual.md]: ☝️ @plaindocs with the really good points
Should we ask folks to check the pre-flight checks and let us know what they get? There are some false alarms in there on RAM, if I recall, but it might help with that and other stuff too

File: user/enterprise/operations-manual.md (hunk 16 above)
Comment (acnagy) [reply user/enterprise/operations-manual.md]: "your installation", or "settings on your installation", maybe?

File: user/enterprise/operations-manual.md (hunk 8 above)
Comment (plaindocs) [reply user/enterprise/operations-manual.md]: Are we talking about changes made trying to troubleshoot, or just "changes to your installation"? Either way, moving into the support bundle makes sense,

**Comment by phuongkt** [comment]
.

**Comment by phuongkt** [comment]
.

**Comment by danishkhan** [comment]
Once this is merged one thing we should look into is adding a link to the operations manual at `8800/support` we can [customize](https://help.replicated.com/docs/kb/developer-resources/branding/) the dashboard. I'll have to test what kind of text we can add, but adding a link to the operations manual and enterprise@travis-ci.com on that page I think will be useful

**Comment by schultyy** [comment]
Hey Hey 👋 

After some internal discussions about the current structure of the document it makes sense to close this PR and start again off a fresh branch. Stay tuned for the new version!
